presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleTtlNodeSelector.java

```diff
@@ -158,7 +160,19 @@ public List<InternalNode> selectRandomNodes(int limit, Set<InternalNode> exclude
 
         Duration estimatedExecutionTimeRemaining = getEstimatedExecutionTimeRemaining();
         List<InternalNode> eligibleNodes = filterNodesByTtl(activeNodes, excludedNodes, ttlInfo, estimatedExecutionTimeRemaining);
-        return selectNodes(limit, new ResettableRandomizedIterator<>(eligibleNodes));
+        List<InternalNode> randomNodes = selectNodes(limit, new ResettableRandomizedIterator<>(eligibleNodes));
+
+        if (randomNodes.isEmpty()) {
+            if (fallbackToSimpleNodeSelection) {
+                log.warn("No nodes available with enough TTL, falling back to simple node selection.");
+                return simpleNodeSelector.selectRandomNodes(limit, excludedNodes);
+            }
```

Comment:
This is to guard against scenarios where a query fails because it does not have nodes with enough TTL to run it. This can happen, for example, if the query was admitted to the cluster but by the time it started executing, the TTLs have reduced (maybe it was incorrectly admitted due to a config issue related to cluster-level TTL, or queries have taken longer to execute than expected and thus the query was in the queue for a long time). In these sort of scenarios, we want to do best effort execution and not fail the query without even trying.